src/checker.rs

```diff
@@ -0,0 +1,88 @@
+use std::error::Error;
+use std::time::Duration;
+
+use reqwest::{ClientBuilder, StatusCode};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum CheckResult {
+    SUCCESS,
+    FAILURE(FailureReason),
+    MISSED,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum FailureReason {
+    Timeout,
+    DnsError(String),
+    ConfigError(String),
+    Error(String),
+}
+
+pub async fn check_domain(domain: String) -> CheckResult {
+    // TODO: Timeout should come from config
+    let timeout = Duration::new(5, 0);
+    let client_result = ClientBuilder::new().timeout(timeout).build();
+    let client = match client_result {
+        Err(e) => {
+            return CheckResult::FAILURE(FailureReason::ConfigError(e.to_string()));
+        }
+        Ok(client) => client,
+    };
```

Comment:
I think there's a more idiomatic way to do this

```suggestion
    let client = client_result.map_err(|e| CheckResult::FAILURE(FailureReason::ConfigError(e.to_string()))?
```

(you'll have to check me on this)

read into the the `?` operator, you'll see it a lot. Basically it unwraps `Result<>` types and early returns if an error comes back.

Reply:
Actually you can probably do this on the client itself

```rs
let client = ClientBuilder::new()
  .timeout(timeout)
  .build();
  .map_err(|e| CheckResult::FAILURE(FailureReason::ConfigError(e.to_string()))?
```

Reply:
The problem with this is that `check_domain` returns a `CheckResult`, which is not the same thing as a `Result` (it has 3 variants to `Result'`'s 2). In general I would strongly prefer using `Result` if at all possible. In this case that could look like
```
enum CheckError {
    Missed,
    Failure(FailureReason),
}
```
`FailureReason` could be exactly as it is now and `CheckResult` could be an alias for `Result<(), CheckError>` (since the `Ok` case contains no information). Alternatively, one could also do
```
enum CheckError {
    Missed,
    Timeout,
    DnsError(String),
    ConfigError(String),
    Error(String),
}
```
and ditch `FailureReason` to remove a level of indirection.

Reply:
I see what you mean. I think you're correct that `CheckResult` should not be a `Result` after all, and that `Missed` should not be a variant of this enum. One simple way to accomplish what you wrote in pseudocode could be to use `Option<CheckResult>` as the type of `result`, with `None` meaning `Missed`.